extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeSignatureUtils.java

```diff
@@ -0,0 +1,150 @@
+    };
+
+    private static final String STS_REGEX = "signatureTimestamp[=:](\\d+)";
+
+    private static final String DEOBF_FUNC_REGEX_START = "(";
+    private static final String DEOBF_FUNC_REGEX_END = "=function\\([a-zA-Z0-9_]+\\)\\{.+?\\})";
+
+    private static final String SIG_DEOBF_HELPER_OBJ_NAME_REGEX = ";([A-Za-z0-9_\\$]{2,})\\...\\(";
+    private static final String SIG_DEOBF_HELPER_OBJ_REGEX_START = "(var ";
+    private static final String SIG_DEOBF_HELPER_OBJ_REGEX_END = "=\\{(?>.|\\n)+?\\}\\};)";
+
+    private YoutubeSignatureUtils() {
+    }
+
+    /**
+     * Get the signature timestamp property of YouTube's base JavaScript file.
+     *
+     * @param javaScriptPlayerCode the complete JavaScript base player code
+     * @return the signature timestamp
+     * @throws ParsingException if the signature timestamp couldn't be extracted
+     */
+    @Nonnull
+    static String getSignatureTimestamp(@Nonnull final String javaScriptPlayerCode)
+            throws ParsingException {
+        try {
+            return Parser.matchGroup1(STS_REGEX, javaScriptPlayerCode);
+        } catch (final ParsingException e) {
+            throw new ParsingException(
+                    "Could not extract signature timestamp from JavaScript code", e);
+        }
+    }
+
+    /**
+     * Get the signature deobfuscation code of YouTube's base JavaScript file.
+     *
+     * @param javaScriptPlayerCode the complete JavaScript base player code
+     * @return the signature deobfuscation code
+     * @throws ParsingException if the signature deobfuscation code couldn't be extracted
+     */
+    @Nonnull
+    static String getDeobfuscationCode(@Nonnull final String javaScriptPlayerCode)
+            throws ParsingException {
+        try {
+            final String deobfuscationFunctionName = getDeobfuscationFunctionName(
+                    javaScriptPlayerCode);
+
+            String deobfuscationFunction;
+            try {
+                deobfuscationFunction = getDeobfuscateFunctionWithLexer(
+                        javaScriptPlayerCode, deobfuscationFunctionName);
+            } catch (final Exception e) {
+                deobfuscationFunction = getDeobfuscateFunctionWithRegex(
+                        javaScriptPlayerCode, deobfuscationFunctionName);
+            }
+
+            // Assert the extracted deobfuscation function is valid
+            JavaScript.compileOrThrow(deobfuscationFunction);
+
+            final String helperObjectName =
+                    Parser.matchGroup1(SIG_DEOBF_HELPER_OBJ_NAME_REGEX, deobfuscationFunction);
+
+            final String helperObject = getHelperObject(javaScriptPlayerCode, helperObjectName);
+
+            final String callerFunction = "function " + DEOBFUSCATION_FUNCTION_NAME
+                    + "(a){return "
+                    + deobfuscationFunctionName
+                    + "(a);}";
+
+            return helperObject + deobfuscationFunction + callerFunction;
```

Comment:
I would add the semicolons here, instead of inside `getHelperObject` and `getDeobfuscateFunctionWithRegex`, for clarity. If the regex in `getHelperObject` needs the `;`, then you can keep it, and add it here anyway, a redundant `;` is valid JS and prevents future programming errors when e.g. changing the regex or the functions.

Reply:
Done, if you meant `getDeobfuscateFunctionWithLexer` and `getDeobfuscateFunctionWithRegex`.